File: app/src/main/java/com/botijasoftware/utils/ArrayBufferShort.java
package com.botijasoftware.utils;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.ShortBuffer;

public class ArrayBufferShort extends ArrayBuffer {
	private ShortBuffer mBuffer;
	private final static int short_size = Short.SIZE/8;

	public ArrayBufferShort(VertexBufferDefinition vbdefinition, int maxelements) {
		super(vbdefinition, maxelements);

		mBuffer = ByteBuffer.allocateDirect(
				vbdefinition.getDataSize() * vbdefinition.mSize * maxelements * short_size)
				.order(ByteOrder.nativeOrder()).asShortBuffer();
	}
	
	public ShortBuffer getBuffer() {
		return mBuffer;
	}
	
	public void put(int index, short value) {
		mBuffer.put(index, value);
	}
	
	public void put(short[] data, int offset) {
		mBuffer.put(data, offset, data.length - offset);
		mBuffer.position(0);
	}
	
	public void put(short[] data) {
		mBuffer.put(data, 0, data.length);
		mBuffer.position(0);
	}

	public void put(byte value) {
		mBuffer.put(value);
	}
	
	public void put(short value) {
		mBuffer.put( value);
	}
	
	public void put(int value) {
		mBuffer.put((short) value);
	}
	
	public void put(float value) {
		mBuffer.put((short)value);
	}
	
	public void position(int i) {
		mBuffer.position(i);
	}
	
	public int position() {
		return mBuffer.position();
	}


}

File: app/src/main/java/com/botijasoftware/utils/ArrayBufferShort.kt
package com.botijasoftware.utils

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.ShortBuffer

class ArrayBufferShort(vbdefinition: VertexBufferDefinition, maxelements: Int) : ArrayBuffer(vbdefinition, maxelements) {
    override val buffer: ShortBuffer

    init {

        buffer = ByteBuffer.allocateDirect(
                vbdefinition.dataSize * vbdefinition.mSize * maxelements * short_size)
                .order(ByteOrder.nativeOrder()).asShortBuffer()
    }

    override fun put(index: Int, value: Short) {
        buffer.put(index, value)
    }

    override fun put(data: ShortArray, offset: Int) {
        buffer.put(data, offset, data.size - offset)
        buffer.position(0)
    }

    override fun put(data: ShortArray) {
        buffer.put(data, 0, data.size)
        buffer.position(0)
    }

    override fun put(value: Byte) {
        buffer.put(value.toShort())
    }

    override fun put(value: Short) {
        buffer.put(value)
    }

    override fun put(value: Int) {
        buffer.put(value.toShort())
    }

    override fun put(value: Float) {
        buffer.put(value.toShort())
    }

    override fun position(i: Int) {
        buffer.position(i)
    }

    override fun position(): Int {
        return buffer.position()
    }

    companion object {
        private val short_size = java.lang.Short.SIZE / 8
    }


}